pkg/compliance/eval/eval.go

```diff
@@ -0,0 +1,462 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package eval
+
+import (
+	"strconv"
+
+	"github.com/alecthomas/participle/lexer"
+	"github.com/alecthomas/repr"
+)
+
+// Evaluatable abstracts part of an expression that can be evaluated for an instance
+type Evaluatable interface {
+	Evaluate(instance *Instance) (interface{}, error)
+}
+
+// Function describes a function callable for an instance
+type Function func(instance *Instance, args ...interface{}) (interface{}, error)
+
+// FunctionMap describes a map of functions
+type FunctionMap map[string]Function
+
+// VarMap describes a map of variables
+type VarMap map[string]interface{}
+
+// Instance for evaluation
+type Instance struct {
+	// Instance functions
+	Functions FunctionMap
+	// Vars defined during evaluation.
+	Vars VarMap
+}
+
+// Iterator abstracts iteration over a set of instances for expression evaluation
+type Iterator interface {
+	Next() (*Instance, error)
+	Done() bool
+}
+
+// InstanceResult captures an Instance along with the passed or failed status for the result
+type InstanceResult struct {
+	Instance *Instance
+	Passed   bool
+}
+
+// Evaluate evaluates an iterable expression for an iterator
+func (e *IterableExpression) Evaluate(it Iterator, global *Instance) (*InstanceResult, error) {
+	if e.IterableComparison == nil {
+		return e.iterate(
+			it,
+			e.Expression,
+			func(instance *Instance, passed bool) bool {
+				// First failure stops the iteration
+				return !passed
+			},
+		)
+	}
+
+	if e.IterableComparison.Fn == nil {
+		return nil, lexer.Errorf(e.Pos, "expecting function for iterable comparison")
+	}
+
+	fn := *e.IterableComparison.Fn
+
+	totalCount := 0
+	passedCount := 0
+	result, err := e.iterate(
+		it,
+		e.IterableComparison.Expression, func(instance *Instance, passed bool) bool {
+			totalCount++
+			if passed {
+				passedCount++
+				if fn == "any" || fn == "none" {
+					return false
+				}
+			} else if fn == "all" {
+				return false
+			}
+			return true
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	passed, err := e.evaluatePassed(global, passedCount, totalCount)
+	if err != nil {
+		return nil, err
+	}
+
+	return &InstanceResult{
+		Instance: result.Instance,
+		Passed:   passed,
+	}, nil
+}
+
+func (e *IterableExpression) evaluatePassed(global *Instance, passedCount, totalCount int) (bool, error) {
+	switch *e.IterableComparison.Fn {
+	case "all":
+		return passedCount == totalCount, nil
+
+	case "none":
+		return passedCount == 0, nil
+
+	case "any":
+		return passedCount != 0, nil
+
+	case "len":
```

Comment:
rename it `count` ?

Reply:
could we also have all the operations declared as consts and represented by an enumeration type?